lib/modules/manager/npm/post-update/pnpm.ts

```diff
@@ -22,13 +23,19 @@ export async function generateLockFile(
   let stderr: string;
   let cmd = 'pnpm';
   try {
-    const pnpmUpdate = upgrades.find(
-      (upgrade) =>
-        upgrade.depType === 'packageManager' && upgrade.depName === 'pnpm'
-    );
+    let pnpmEngine;
+    const rootPackageJson = upath.join(cwd, 'package.json');
+    const content = await readFile(rootPackageJson, 'utf8');
+    if (content) {
+      const packageJson: NpmPackage = JSON.parse(content);
+      const engines = packageJson?.engines;
+      pnpmEngine = engines['pnpm'];
+    }
     const pnpmToolConstraint: ToolConstraint = {
       toolName: 'pnpm',
-      constraint: pnpmUpdate ? pnpmUpdate.newValue : config.constraints?.pnpm,
+      constraint: config.constraints?.pnpm
+        ? config.constraints.pnpm
+        : pnpmEngine,
```

Comment:
```suggestion
      constraint: config.constraints?.pnpm?? await getPnpmContraint(),
```

Make reading pnpm contraints lazy, also try the new `packageManager` field.

- https://nodejs.org/api/packages.html#packages_packagemanager
- https://nodejs.org/api/corepack.html

Reply:
so the packageManager is the new field for deciding, now which one takes precedence? 
i cant find any connection between the two online, 
also
![image](https://user-images.githubusercontent.com/42116482/161936981-b54a5b05-5eb8-4b27-9542-d5a410c2abf0.png)
should i consider the flag? or just do the constraint anyway?

Reply:
`packageManager` is the new way. I think `engines` should have preference for now as it's more commonly used.
@rarkins WDYT?

Reply:
well, if everybody is using engine, and only 1 guy is using packageManager, we can check for packageManager first since its the new field, if people dont have it, then we fall back to engine.
that makes more sense to me.

the only reason why i would agree with you is because we could keep backward compatibility if the engines precedes the pacakgeManager.

ill let you and @rarkins decide though :D

Reply:
You are right, in future corepack would probably enabled by default, so `packageManager` should have preference 👍